x-pack/plugins/endpoint/public/embeddables/resolver/types.ts

```diff
@@ -130,13 +131,60 @@ export type CameraState = {
     }
 );
 
+/**
+ * This represents all the raw data (sans statistics, metadata, etc.)
+ * about a particular subject's related events
+ */
+export interface RelatedEventDataEntry {
+  relatedEvents: Array<{
+    relatedEvent: ResolverEvent;
+    relatedEventType: ReturnType<typeof eventType>;
+  }>;
+}
+
+/**
+ * Represents the status of the request for related event data, which will be either the data,
+ * a value indicating that it's still waiting for the data or an Error indicating the data can't be retrieved as expected
+ */
+export type RelatedEventDataResults =
```

Comment:
Going through this one last time - I think this lends itself to a metaprogramming style typescript is supposed to protect against. Having 3 different possible types is often times about as good as an `any` in certain situations and forces you to have to narrow down the types every time it's passed (Like you do [here](https://github.com/elastic/kibana/pull/64121/files#diff-b18117271821fdd985eed1e1c4fd749fR375)). Using something more certain and perhaps a different type to keep track of the fetch status would be a lot more straightforward for the control flow of the functions and make it easier to track for developers. You're already doing something similar [here](https://github.com/elastic/kibana/pull/64121/files#diff-21b3c8993d0c513e9fd1152ef197f46eR125) with the checking of the fetch status for a loading field, I'm not sure what the benefit is of having that functionality coupled with the data type itself, I think it'd be clearer to have them separate. I realize this might require a lot of refactoring so I'm not gonna block but just my thoughts on it

Reply:
Thanks @dplumlee . I think every reviewer expressed that they aren't really comfortable packing this many concepts into  one value, so I'll log an issue to refactor this when we rewrite it to use the new related info as it get mainlined in with the  tree results.

Reply:
I'd point out that separating these concepts from each other in the type as described here _doesn't obviate the need to handle that complexity in the app_. You still have to _handle_ the cases where results haven't arrived yet, or there was some problem fetching them. AFAICT, you still have to handle checking those concepts in the view layer no matter how you send them down: Whether you pack them up and send them in one channel or split them into 3 props didn't seem to be that different when I considered it, but as it's purely a point of style I'll conform to what clearly seems to be the consensus here and crack them apart when we refactor.

Reply:
Also, differentiating behavior based on a `typeof` or `instanceof` doesn't qualify as metaprogramming (at as I understand it).

Reply:
@dplumlee  Marking this as Resolved with a note for the upcoming refactor: https://github.com/elastic/endpoint-app-team/issues/425

Reply:
yeah metaprogramming might not be the perfect term for it, I was struggling to come up with something exact (deterministic maybe?), but I think you understood what I was going for. And I do agree it's largely a matter of style but I think that the separate channels would be clearer for flow. Thanks

Reply:
@dplumlee "This type is over-articulated for its intended purpose" might be a more apt way to describe what bothers you about that.

Reply:
:straight_ruler:  Don't: Multiplex results in complex types Do: Separate data/results from statii into their own values/channels